File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/Control.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.graphics.Brush;
import com.hello.uxml.tools.framework.graphics.Color;
import com.hello.uxml.tools.framework.graphics.UISurface;

import java.util.EnumSet;

/**
 * Provides control level properties such as FontName, FontSize,
 * tabbing support, etc.
 *
 * @author ferhat@(Ferhat Buyukkokten)
 */
public class Control extends UIElement {

  /**
   * Chrome element tree.
   */
  protected UIElement chromeTree;
  public static final String DEFAULT_FONT_NAME = "Arial";
  public static final double DEFAULT_FONT_SIZE = 10.0;
  public static final Boolean DEFAULT_FONT_BOLD = false;
  /** cached padding */
  protected Margin cachedPadding = Margin.EMPTY;

  /** FontName Property Definition */
  public static PropertyDefinition fontNamePropDef = PropertySystem.register("FontName",
      String.class,
      Control.class,
      new PropertyData(DEFAULT_FONT_NAME, EnumSet.of(PropertyFlags.Resize,
          PropertyFlags.Redraw, PropertyFlags.Inherit)));

  /** FontSize Property Definition */
  public static PropertyDefinition fontSizePropDef = PropertySystem.register("FontSize",
      Double.class,
      Control.class,
      new PropertyData(DEFAULT_FONT_SIZE, EnumSet.of(PropertyFlags.Resize,
          PropertyFlags.Redraw, PropertyFlags.Inherit)));

  /** FontBold Property Definition */
  public static PropertyDefinition fontBoldPropDef = PropertySystem.register("FontBold",
      Boolean.class,
      Control.class,
      new PropertyData(DEFAULT_FONT_BOLD, EnumSet.of(PropertyFlags.Resize,
          PropertyFlags.Redraw, PropertyFlags.Inherit)));

  /** TextColor Property Definition */
  public static PropertyDefinition textColorPropDef = PropertySystem.register("TextColor",
      Color.class,
      Control.class,
      new PropertyData(Color.BLACK, EnumSet.of(PropertyFlags.Redraw, PropertyFlags.Inherit)));

  /** Background property definition */
  public static PropertyDefinition backgroundPropDef = PropertySystem.register("Background",
      Brush.class, Control.class, new PropertyData(null, null, EnumSet.of(
          PropertyFlags.Redraw)));

  /** Padding Property Definition */
  public static PropertyDefinition paddingPropDef = PropertySystem.register("Padding",
      Margin.class, Control.class,
      new PropertyData(Margin.EMPTY, new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
          ((Control) e.getSource()).cachedPadding = ((Margin) e.getNewValue());
          ((Control) e.getSource()).invalidateSize();
        }}));

  /** Border radius Property Definition */
  public static PropertyDefinition borderRadiusPropDef = PropertySystem.register("BorderRadius",
      BorderRadius.class, Control.class,
      new PropertyData(BorderRadius.EMPTY, EnumSet.of(PropertyFlags.Redraw)));

  /** Prompt message property definition */
  public static PropertyDefinition promptMessagePropDef = PropertySystem.register("PromptMessage",
      String.class,
      Control.class,
      new PropertyData("", EnumSet.of(PropertyFlags.Redraw, PropertyFlags.Localizable)));

  /** Validation error message property definition */
  public static PropertyDefinition errorMessagePropDef = PropertySystem.register("ErrorMessage",
      String.class,
      Control.class,
      new PropertyData("", EnumSet.of(PropertyFlags.None, PropertyFlags.Localizable)));

  /** Chrome property definition */
  public static PropertyDefinition chromePropDef = PropertySystem.register("Chrome",
      Chrome.class, Control.class, new PropertyData(null, EnumSet.of(PropertyFlags.None),
          new PropertyChangeListener() {
            @Override
            public void propertyChanged(PropertyChangedEvent e) {
              ((Control) e.getSource()).onChromeChanged((Chrome) e.getNewValue());
            }}));

  /**
   * Sets or returns font name.
   */
  public String getFontName() {
    return ((String) getProperty(fontNamePropDef));
  }

  public void setFontName(String name) {
    setProperty(fontNamePropDef, name);
  }

  /**
   * Sets or returns prompt message.
   */
  public String getPromptMessage() {
    return ((String) getProperty(promptMessagePropDef));
  }

  public void setPromptMessage(String name) {
    setProperty(promptMessagePropDef, name);
  }

  /**
   * Sets or returns validation error message.
   */
  public String getErrorMessage() {
    return ((String) getProperty(errorMessagePropDef));
  }

  public void setErrorMessage(String name) {
    setProperty(errorMessagePropDef, name);
  }

  /**
   * Sets or returns font size.
   */
  public double getFontSize() {
    return ((Double) getProperty(fontSizePropDef)).doubleValue();
  }

  public void setFontSize(double size) {
    setProperty(fontSizePropDef, size);
  }

  /**
   * Sets or returns font bold.
   */
  public boolean getFontBold() {
    return ((Boolean) getProperty(fontBoldPropDef));
  }

  public void setFontBold(boolean bold) {
    setProperty(fontBoldPropDef, bold);
  }

  /**
   * Sets or returns text color.
   */
  public Color getTextColor() {
    return (Color) getProperty(textColorPropDef);
  }

  public void setTextColor(Color color) {
    setProperty(textColorPropDef, color);
  }

  /** Gets or sets background fill brush */
  public Brush getBackground() {
    return (Brush) getProperty(backgroundPropDef);
  }

  public void setBackground(Brush value) {
    setProperty(backgroundPropDef, value);
  }

  /**
   * Sets/returns border radius.
   */
  public void setBorderRadius(BorderRadius value) {
    setProperty(borderRadiusPropDef, value);
  }

  public BorderRadius getBorderRadius() {
    return (BorderRadius) getProperty(borderRadiusPropDef);
  }

  /**
   * Sets/returns padding.
   */
  public void setPadding(Margin value) {
    setProperty(paddingPropDef, value);
  }

  public Margin getPadding() {
    return cachedPadding;
  }

  @Override
  protected void onRedraw(UISurface surface) {
    Brush backgroundBrush = getBackground();
    if (backgroundBrush != null) {
      surface.drawRect(backgroundBrush, null, new Rectangle(0, 0, getLayoutRect().width,
          getLayoutRect().height));
    }
  }

  protected void onChromeChanged(Chrome chrome) {
    // Remove old chrome tree
    if (chromeTree != null) {
      removeRawChild(chromeTree);
      chromeTree = null;
    }
    if (chrome != null) {
      chromeTree = chrome.apply(this);
      if (chromeTree != null) {
        addRawChild(chromeTree);
      }
    }
  }

  @Override
  public void initSurface(UISurface parentSurface) {
    // Initialize chrome before surface initialization
    if (getChrome() == null) {
      Chrome chrome = (Chrome) findResource(getClass());
      if (chrome != null) {
        setChrome(chrome);
      }
    }
    super.initSurface(parentSurface);
  }

  /**
   * Sets/returns chrome of control.
   */
  public Chrome getChrome() {
    return (Chrome) getProperty(chromePropDef);
  }

  public void setChrome(Chrome chrome) {
    setProperty(chromePropDef, chrome);
  }

  @Override
  protected int getRawChildCount() {
    return (chromeTree == null) ? 0 : 1;
  }

  @Override
  protected UIElement getRawChild(int index) {
    return chromeTree;
  }

  @Override
  protected void onMeasure(double availableWidth, double availableHeight) {
    double maxWidth = 0;
    double maxHeight = 0;
    if (chromeTree != null) {
      chromeTree.measure(availableWidth, availableHeight);
      maxWidth = chromeTree.getMeasuredWidth();
      maxHeight = chromeTree.getMeasuredHeight();
    } else if (getRawChildCount() != 0) {
      UIElement child = getRawChild(0);
      child.measure(availableWidth, availableHeight);
      maxWidth = child.getMeasuredWidth();
      maxHeight = child.getMeasuredHeight();
    }
    setMeasuredDimension(maxWidth, maxHeight);
  }

  @Override
  protected void onLayout(Rectangle targetRect) {
    if (chromeTree != null) {
      chromeTree.layout(targetRect);
    } else if (getRawChildCount() != 0) {
      getRawChild(0).layout(targetRect);
    }
  }
}

File: framework/lib/control.dart
part of uxml;

/**
 * Provides control level properties such as FontName, FontSize,
 * tabbing support, etc.
 *
 * @author:ferhat@ (Ferhat Buyukkokten)
 */
class Control extends UIElement {

  static const String DEFAULT_FONT_NAME = "Helvetica";
  static const num DEFAULT_FONT_SIZE = 10.0;
  static const bool DEFAULT_FONT_BOLD = false;
  static Color DEFAULT_TEXT_COLOR;

  /** Border radius property definition */
  static PropertyDefinition borderRadiusProperty;
  /** FontName property definition */
  static PropertyDefinition fontNameProperty;
  static PropertyDefinition backgroundProperty;
  /** FontSize property definition */
  static PropertyDefinition fontSizeProperty;
  /** FontBold property definition */
  static PropertyDefinition fontBoldProperty;
  /** Textcolor property definition */
  static PropertyDefinition textColorProperty;
  /**
   * LinkColor property definition, specifies color used for inline
   * hyperlinks
   */
  static PropertyDefinition linkColorProperty;
  /** Chrome property definition */
  static PropertyDefinition chromeProperty;
  /** Padding Property Definition */
  static PropertyDefinition paddingProperty;
  /** Prompt message property definition */
  static PropertyDefinition promptMessageProperty;
  /** Validation error message property definition */
  static PropertyDefinition errorMessageProperty;
  static ElementDef controlElementDef;

  /** Cached padding value */
  Margin _cachedPadding;

  /**
   * Holds ui tree provided by chrome.
   */
  UIElement chromeTree = null;

  // chromeDirty flag is used to delay generating the chromeTree until
  // a visible surface is initialized.
  // If user's accesses DOM tree before surface initialization we force
  // chromeTree creation since code will need to find elements by ids.
  bool _chromeDirty = false;

  /**
   * Constructor
   */
  Control() : super() {
    _cachedPadding = Margin.EMPTY;
  }

  /** @see UIElement.initSurface. */
  void initSurface(UISurface parentSurface, [int index = -1]) {
    // Initialize chrome before surface initialization
    if (chrome == null) {
      Chrome res = findResource(getDefinition());
      if (res != null) {
        chrome = res;
      }
    }
    if (_chromeDirty) {
      _generateChromeTree();
      applyEffects();
    }
    super.initSurface(parentSurface, index);
    if (borderRadius != null && _hostSurface != null) {
      _hostSurface.setBorderRadius(borderRadius);
    }
  }

  void _generateChromeTree() {
    _chromeDirty = false;
    Object prevController = null;
    if (overridesProperty(Controller.controllerProperty)) {
      prevController = Controller.getTargetController(this);
    }
    if (chrome != null) {
      chromeTree = chrome.applyToTarget(this);
      if (chromeTree != null) {
        insertRawChild(chromeTree, -1);
      }
      if (overridesProperty(Controller.controllerProperty)) {
        Controller newController = Controller.getTargetController(this);
        if (prevController != newController) {
          newController.initCompleted();
        }
      }
    }
    onChromeTreeReady();
  }

  /**
   * Provides override to access chromeTree after lazy initialization.
   */
  void onChromeTreeReady() {
  }

  int getRawChildCount() => (chromeTree == null) ? 0 : 1;

  UIElement getRawChild(int index) => chromeTree;

  /**
  * Sets or returns background brush.
  */
  Brush get background => getProperty(backgroundProperty);

  set background(Brush brush) {
    setProperty(backgroundProperty, brush);
  }

  /**
   * Sets or returns border radius.
   */
  BorderRadius get borderRadius => getProperty(borderRadiusProperty);

  set borderRadius(BorderRadius value) {
    setProperty(borderRadiusProperty, value);
  }

  /** Gets or sets the padding of a border */
  Margin get padding => _cachedPadding;

  set padding(Margin value) {
    setProperty(paddingProperty, value);
  }

  /**
  * Sets or returns font name.
  */
  String get fontName => getProperty(fontNameProperty);

  set fontName(String familyName) {
    setProperty(fontNameProperty, familyName);
  }

  /**
   * Sets or returns font size.
   */
  num get fontSize => getProperty(fontSizeProperty);

  set fontSize(num size) {
    setProperty(fontSizeProperty, size);
  }

  /**
   * Sets or returns font bold.
   */
  bool get fontBold => getProperty(fontBoldProperty);

  set fontBold(bool bold) {
    setProperty(fontBoldProperty, bold);
  }

  /**
   * Sets or returns text color.
   */
  Color get textColor => getProperty(textColorProperty);

  set textColor(Color color) {
    setProperty(textColorProperty, color);
  }

  /**
  * Sets or returns chrome of control.
  */
  Chrome get chrome => getProperty(chromeProperty);

  set chrome(Chrome value) {
    setProperty(chromeProperty, value);
  }

  /**
   * Sets or returns prompt message.
   */
  String get promptMessage => getProperty(promptMessageProperty);

  set promptMessage(String message) {
    setProperty(promptMessageProperty, message);
  }

  /**
   * Sets or returns validation error message.
   */
  String get errorMessage => getProperty(errorMessageProperty);

  set errorMessage(String message) {
    setProperty(errorMessageProperty, message);
  }

  /** Overrides UIElement.onRedraw. */
  void onRedraw(UISurface surface) {
    if (_hostSurface != null) {
      _hostSurface.setBackground(background);
      _hostSurface.setBorderRadius(borderRadius);
    }
  }

  /** Overrides UIElement.onMeasure to return size of element. */
  bool onMeasure(num availableWidth, num availableHeight) {
    if (_chromeDirty) {
      _generateChromeTree();
    }
    num maxWidth = 0.0;
    num maxHeight = 0.0;
    num pw = padding.left + padding.right;
    num ph = padding.top + padding.bottom;
    if (chromeTree != null) {
      chromeTree.measure(availableWidth - pw, availableHeight - ph);
      maxWidth = chromeTree.measuredWidth;
      maxHeight = chromeTree.measuredHeight;
    }
    setMeasuredDimension(maxWidth + pw, maxHeight + ph);
    return false;
  }

  /** Overrides UIElement.onLayout. */
  void onLayout(num targetX, num targetY, num targetWidth, num targetHeight) {
    if (chromeTree != null) {
      chromeTree.layout(_cachedPadding.left, _cachedPadding.top,
          targetWidth - _cachedPadding.left - _cachedPadding.right,
          targetHeight - _cachedPadding.top - _cachedPadding.bottom);
    }
  }

  UIElement getElement(String elementId) {
    if (_chromeDirty) {
      _generateChromeTree();
    }
    UIElement element = super.getElement(elementId);
    if (element == null && chromeTree != null) {
      return chromeTree.getElement(elementId);
    }
    return element;
  }

  void onChromeChanged(Chrome oldChrome, Chrome newChrome) {
    // Remove old chrome tree
    if (chromeTree != null) {
      removeRawChild(chromeTree);

      // Also remove effects added by the old chrome.
      for (int i = 0; i < oldChrome.effects.length; i++) {
        Effect oldEffect = oldChrome.effects[i];
        for (int e = 0; e < effects.length; e++) {
          if (oldEffect.id == effects[e].id) {
            effects.removeAt(effects.indexOf(effects[e]));
            break;
          }
        }
      }
      chromeTree = null;
    }
    if (_hostSurface == null) {
      _chromeDirty = true;
    } else {
      _generateChromeTree();
      applyEffects();
    }
  }
  /**
   * Override stub for classes.
   */
  void onFontChanged() {
  }

  static void _borderRadiusChangedHandler(Object target,
                                          Object property,
                                          Object oldValue,
                                          Object newValue) {
    UIElement element = target;
    UISurface surface = element._hostSurface;
    if (surface != null) {
      BorderRadius border = newValue;
      surface.setBorderRadius(border);
    }
  }

  static void _paddingChangedHandler(Object target, Object property,
      Object oldValue, Object newValue) {
    Control control = target;
    control._cachedPadding = (newValue == PropertyDefaults.NO_DEFAULT) ? null :
        newValue;
  }

  static void _fontChangedHandler(Object target,
                                  Object property,
                                  Object oldValue,
                                  Object newValue) {
    if (target is Control) {
      Control control = target;
      control.onFontChanged();
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => controlElementDef;

  /** Registers component. */
  static void registerControl() {
    DEFAULT_TEXT_COLOR = Color.fromRGB(0);
    borderRadiusProperty = ElementRegistry.registerProperty("borderRadius",
        PropertyType.BORDERRADIUS, PropertyFlags.REDRAW,
        _borderRadiusChangedHandler, null);
    fontNameProperty = ElementRegistry.registerProperty("fontName",
        PropertyType.STRING, PropertyFlags.RESIZE | PropertyFlags.INHERIT,
        _fontChangedHandler, DEFAULT_FONT_NAME);
    backgroundProperty = ElementRegistry.registerProperty("background",
        PropertyType.BRUSH, PropertyFlags.REDRAW, null, null);
    fontSizeProperty = ElementRegistry.registerProperty("fontSize",
        PropertyType.NUMBER, PropertyFlags.RESIZE | PropertyFlags.INHERIT,
        _fontChangedHandler, DEFAULT_FONT_SIZE);
    fontBoldProperty = ElementRegistry.registerProperty("fontBold",
        PropertyType.BOOL, PropertyFlags.RESIZE | PropertyFlags.INHERIT,
        _fontChangedHandler, DEFAULT_FONT_BOLD);
    textColorProperty = ElementRegistry.registerProperty("textColor",
        PropertyType.COLOR, PropertyFlags.INHERIT,
        _fontChangedHandler, Color.fromRGB(0));
    linkColorProperty = ElementRegistry.registerProperty("linkColor",
        PropertyType.COLOR, PropertyFlags.INHERIT,
      _fontChangedHandler, Color.fromRGB(0x0000CC));
    chromeProperty = ElementRegistry.registerProperty("chrome",
        PropertyType.CHROME, PropertyFlags.NONE,
        (UIElement target, PropertyDefinition property,
            Object oldVal, Object newVal) {
          if (target is Control) {
            Control control = target;
            control.onChromeChanged(oldVal, newVal);
          }
        }, null);

    paddingProperty = ElementRegistry.registerProperty("padding",
        PropertyType.MARGIN, PropertyFlags.RESIZE, _paddingChangedHandler,
        Margin.EMPTY);
    promptMessageProperty = ElementRegistry.registerProperty("promptMessage",
        PropertyType.STRING, PropertyFlags.NONE, null, null);
    errorMessageProperty = ElementRegistry.registerProperty("errorMessage",
        PropertyType.STRING, PropertyFlags.NONE, null, null);
    controlElementDef = ElementRegistry.register("Control",
        UIElement.elementDef,
        [backgroundProperty, borderRadiusProperty, fontNameProperty,
        fontSizeProperty, fontBoldProperty, textColorProperty, chromeProperty,
        paddingProperty, promptMessageProperty, errorMessageProperty], null);
  }
}
